### PR TITLE
chore: Initialize copy array from previous values in `array_set`

### LINKED ISSUE #2016: Add rationale as to why we create a new array with zero values
### Problem

See : https://github.com/noir-lang/noir/pull/1969#discussion_r1272297218

When we do an array_set, we create a new array with zero values, then copy the old array into this new array, overwriting the zeroes.

Question was referring to why we don't just copy the new values into the array initially

### Happy Case

.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Initialise array with values it is copied from during array_set

## Problem\*
Arrays were initialised to 0 and then copied from.

Resolves #2016

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs [hunk 1]
```diff
@@ -602,19 +602,19 @@ impl Context {
         let result_array_id = result_id.to_usize() as u32;
         let result_block_id = BlockId(result_array_id);
 
-        // Initialize the new array with zero values
-        self.initialize_array(result_block_id, len, None)?;
-
-        // Copy the values from the old array into the newly created zeroed array
+        // Initialize the new array with the values from the old array
+        let mut init_values = Vec::new();
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:606]: We can use `try_vecmap` here

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:606]: Not really, the error handling will get too cumbersome

Comment (jfecher) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:606]: No, the error handling is the same, you can still use `?`

Comment (TomAFrench) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs]: I've switched this over to use `try_vecmap`.

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs [hunk 2]
```diff
@@ -602,19 +602,18 @@ impl Context {
...
             let index = AcirValue::Var(
                 self.acir_context.add_constant(FieldElement::from(i as u128)),
                 AcirType::NumericType(NumericType::NativeField),
             );
             let var = index.into_var()?;
             let read = self.acir_context.read_from_memory(block_id, &var)?;
-            self.acir_context.write_to_memory(result_block_id, &var, &read)?;
-        }
+            let acir_value = AcirValue::Var(read, AcirType::NumericType(NumericType::NativeField));
+            Ok(acir_value)
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:614]:
```suggestion
            Ok(AcirValue::Var(read, AcirType::NumericType(NumericType::NativeField))_
```

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs [hunk 3]
```diff
@@ -602,19 +602,17 @@ impl Context {
...
+        let init_values = try_vecmap(0..len, |i| {
             let index = AcirValue::Var(
                 self.acir_context.add_constant(FieldElement::from(i as u128)),
                 AcirType::NumericType(NumericType::NativeField),
             );
             let var = index.into_var()?;
             let read = self.acir_context.read_from_memory(block_id, &var)?;
-            self.acir_context.write_to_memory(result_block_id, &var, &read)?;
-        }
+            Ok(AcirValue::Var(read, AcirType::NumericType(NumericType::NativeField))_
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:613]:
```suggestion
            Ok(AcirValue::Var(read, AcirType::NumericType(NumericType::NativeField)))
```

**Review by jfecher — DISMISSED** [review]

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**